components/raftstore/src/store/worker/region.rs

```diff
@@ -620,6 +635,46 @@ where
 
         Ok(())
     }
+
+    fn pre_apply_snapshot(&self, task: &Task<EK::Snapshot>) -> Result<()> {
+        let (region_id, abort, peer_id) = match task {
+            Task::Apply {
+                region_id,
+                status,
+                peer_id,
+            } => (region_id, status.clone(), peer_id),
+            _ => panic!("invalid apply snapshot task"),
+        };
+
+        let region_state = self.get_region_state(*region_id)?;
+        let apply_state = self.get_apply_state(*region_id)?;
+
+        let timer = Instant::now();
+        check_abort(&abort)?;
+
+        let term = apply_state.get_truncated_state().get_term();
+        let idx = apply_state.get_truncated_state().get_index();
+        let snap_key = SnapKey::new(*region_id, term, idx);
+        let s = box_try!(self.mgr.get_snapshot_for_applying(&snap_key));
+        if !s.exists() {
+            return Err(box_err!("missing snapshot file {}", s.path()));
+        }
+        check_abort(&abort)?;
+        self.coprocessor_host.pre_apply_snapshot(
+            region_state.get_region(),
+            *peer_id,
+            &snap_key,
+            Some(&s),
+        );
+        info!(
+            "pre apply snapshot";
+            "region_id" => region_id,
+            "peer_id" => peer_id,
+            "state" => ?apply_state,
+            "time_takes" => ?timer.saturating_elapsed(),
+        );
```

Comment:
does it need to be info!?
I think maybe debug! and add a pre-snapshot duration metrics makes more sense.

Reply:
If tiflash has its own metrics, then I think we don't need another metrics.